File: backend/src/main/java/com/tl/models/application/game/ws_messages/messages/StateChangedMessage.java
package com.tl.models.application.game.ws_messages.messages;

import com.fasterxml.jackson.annotation.JsonProperty;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.tl.models.application.game.states.GameStateIdentifier;
import com.tl.models.application.game.ws_messages.Message;
import com.tl.models.application.game.ws_messages.message_type.MessageType;
import lombok.Data;
import lombok.EqualsAndHashCode;
import lombok.SneakyThrows;

public class StateChangedMessage<T> extends Message {
    @JsonProperty
    private GameStateIdentifier next;
    @JsonProperty
    private T data;

    public StateChangedMessage(GameStateIdentifier next, T data) {
        super(MessageType.StateChanged);
        this.next = next;
        this.data = data;
    }

    @SneakyThrows
    @Override
    public String serialize() {
        return new ObjectMapper().writeValueAsString(this);
    }
}

File: frontend/dog-web/src/app/models/websockets/state-changed-message.ts
import { GameState } from "../game-state";
import { MessageTypeMapper } from "./dto/message-type.dto";
import { StateChangedMessageDTO } from "./dto/state-changed-message.dto";
import { Message } from "./message";
import { MessageType } from "./message-type";

export class StateChangedMessage<TData> extends Message {
    constructor(
        type: MessageType,
        public data: TData,
        public next: GameState
    ) {
        super(type);
    }

    public static fromApi<TResult>({type, data, next}: StateChangedMessageDTO<TResult>): StateChangedMessage<TResult> {
        return new StateChangedMessage(MessageTypeMapper.fromApi(type), data, next as GameState);
    }
}